README.md
git brws
========
[![crates.io][crate-badge]][crate]
[![Linux and macOS CI][travis-badge]][travis-ci]
[![Windows CI][appveyor-badge]][appveyor]
[![Coverage][codecov-badge]][codecov]

`git brws` is a command line tool to open a repository, file, commit, diff, tag, blame, pull request,
issue or project's website in your web browser from command line. 'brws' is an abbreviation of 'browse'.

Features:

- Opens a page of
  - Repository ([example](https://github.com/rhysd/git-brws))
  - File ([example](https://github.com/rhysd/git-brws/blob/master/Cargo.toml))
  - Commit ([example](https://github.com/rhysd/git-brws/commit/60024ab1280f9f10423b22bc708f3f6ef97db6b5))
  - Diff ([example](https://github.com/rhysd/git-brws/compare/e3c18d0d50252112d37bde97061370204b3cdab7..60024ab1280f9f10423b22bc708f3f6ef97db6b5), [example](https://github.com/rhysd/git-brws/compare/e3c18d0d50252112d37bde97061370204b3cdab7...60024ab1280f9f10423b22bc708f3f6ef97db6b5))
  - Tag ([example](https://github.com/rhysd/git-brws/tree/0.10.0))
  - Pull request (only for GitHub and GitHub Enterprise) ([example](https://github.com/rust-lang/rust.vim/pull/290))
  - Issue ([example](https://github.com/rhysd/git-brws/issues/8))
  - Website ([example](https://rhysd.github.io/git-brws/))
    - Homepage of the repository for GitHub or GitHub Enterprise, [GitLab pages][gitlab-pages], [Bitbucket Cloud][bitbucket-cloud]
  - Blame ([example](https://github.com/rhysd/git-brws/blame/9ab093f6720c2c2fe8375408f7f3ab40a3d3337a/src/service.rs))
- Supports following services
  - [GitHub](https://github.com)
  - [Bitbucket](https://bitbucket.org)
  - [GitHub Enterprise](https://enterprise.github.com/home)
  - [GitLab](https://about.gitlab.com/)
  - [Azure DevOps](https://azure.microsoft.com/services/devops/)
- Prefers commit-specific page URL (permlink)
- Available on Linux, macOS and Windows

## Installation

`git brws` is available on Linux, macOS and Windows.

### With [Homebrew](https://brew.sh/)

```
$ brew tap "rhysd/git-brws" "https://github.com/rhysd/git-brws"
$ brew install git-brws
```

It installs `git-brws` to `/usr/local/bin` and `git-brws.1` to `/usr/local/share/man/man1`.
This is recommended way for installation on macOS since updating to the new version is easy.

### On [Arch Linux](https://www.archlinux.org/)

You can install `git-brws` via the [AUR package](https://aur.archlinux.org/packages/git-brws/):

```
git clone https://aur.archlinux.org/git-brws.git
cd git-brws
makepkg -si
```

### Download from [release page][]

Pre-built binary executables are available at [release page][] for following targets:

- Linux 64bit (x86_64-unknown-linux-gnu)
- macOS 64bit (x86_64-apple-darwin)
- Windows 64bit (x86_64-pc-windows-msvc)
- Linux 32bit (i686-unknown-linux-gnu)
- Windows 32bit (i686-pc-windows-msvc)

Download and unarchive the binary then put the executable in `$PATH`.

Manpage for `man` command is available. Please find `git-brws.1` in the unarchived directory or download
it from [here](https://raw.githubusercontent.com/rhysd/git-brws/master/git-brws.1) and copy it to the
`man` directory in your system (e.g. `/usr/local/share/man/man1/`).

Note: `git brws --help` only works when it is installed. `git brws -h` is always available.

### With [cargo](https://crates.io/)

```
$ cargo install git-brws
```

It requires Rust toolchain to build git-brws from sources.

## Usage

```
Usage: git brws [Options] {Args}

Options:
    -r, --repo REPO     Shorthand format (repo, user/repo, host/user/repo) or
                        Git URL you want to see. When only repo name is
                        specified, most popular repository will be searched
                        from GitHub
    -b, --branch BRANCH Branch name to browse
    -d, --dir PATH      Directory path to the repository. Default value is
                        current working directory
    -R, --remote REMOTE Remote name (e.g. origin). Default value is a remote
                        the current branch is tracking. If current branch
                        tracks no branch, it falls back to 'origin'
    -u, --url           Output URL to stdout instead of opening in browser
    -p, --pr            Open pull request page instead of repository page. If
                        not existing, open 'Create Pull Request' page
    -w, --website       Open website page instead of repository page (homepage
                        URL for GitHub, GitLab pages, Bitbucket Cloud)
    -B, --blame         Open blame page instead of repository page. File path
                        to blame must be passed also.
    -c, --current-branch
                        Open the current branch instead of default branch
    -h, --help          Print this help
    -v, --version       Show version
```

### Open a repository page

- Repository at current directory

```
$ git brws
```

- `develop` branch

```
$ git brws -b develop
```

- `origin` remote of `develop` branch

```
$ git brws -r origin -b develop
```

- @rhysd's 'Shiba' repository

```
$ git brws -r rhysd/Shiba
```

- Most popular `react` repository on GitHub

```
$ git brws -r react
```

It will open https://github.com/facebook/react.

Note: When only repository name is specified for `-r` option, `git-brws` searches GitHub with query
`{repo} in:name` and opens the best-matched repository page.

- Specify directory of repository

```
$ git brws -d ~/.go/src/github.com/rhysd/dot-github
```

### Open a file page

- File of current branch of current remote

```
$ git brws ./some/file.txt
```

- Line of the file

```
$ git brws ./some/file.txt#L123
```

Note: The `L` can be omit.

- Range from line to line of the file

```
$ git brws ./some/file.txt#L123-L126
```

Note: The `L` can be omit.

### Open a commit page

- `HEAD` page of current repository

```
$ git brws HEAD
```

### Open a tag page

- `0.10.0` tag page of current repository

```
$ git brws 0.10.0
```

### Open a diff page between commits

- Diff between `HEAD` and `HEAD~3`

```
$ git brws HEAD~3..HEAD
```

- Diff between `113079b` and `60024ab`

```
$ git brws 60024ab..113079b
```

### Open a diff page from specific commit and its merge base

In addition to `..`, diff with `...` is supported.

- Diff between `branchB` and the merge base commit from `branchB` into `branchA`

```
$ git brws branchA...branchB
```

If you don't know the difference between `..` and `...`, please read `git diff --help`.

Note: Only GitHub and GitHub Enterprise support `...`. For GitLab, only `...` is available.

### Open a pull request page

- Pull request page of current branch of current repository

```
$ git brws --pr
```

- Pull request page of specific branch of specific repository

```
# Specify my forked repository
$ git brws --pr --repo rhysd/rust.vim -b async-contextual-keyword

# Or specify original repository
$ git brws --pr --repo rust-lang/rust.vim -b async-contextual-keyword
```

Note: Currently only GitHub and GitHub Enterprise are supported.

Note: If you have created multiple pull requests at the same repository with the same branch name,
the command may not open a pull request page you want.

Note: When a pull request page for current branch is not existing yet, it opens 'Create Pull Request'
page instead.

### Open a website for the repository

```
# Website for current repository
$ git brws --website
```

With `--repo` option, arbitrary repository's website can be opened.

```
# Opens https://reactjs.org
$ git brws --website --repo react
```

It opens a website for the repository.

- For GitHub, URL for 'homepage' configuration of the repository if it's set. Otherwise
  `https://{user}.github.io/{repo}`
- For GitHub Enterprise, `https://pages.{host}/{user}/{repo}` or `https://{host}/pages/{user}/{repo}`
  depending on your GitHub Enterprise configuration of subdomain isolation
- For GitLab, [GitLab Pages][gitlab-pages]
- For Bitbucket, [Bitbucket Cloud][bitbucket-cloud]

### Open an issue page

- Issue #8

```
$ git brws '#8'
```

Note: `#` is usually used for a line comment in major shells. Please quote the argument

### Open a blame page

- Specific file

```
$ git brws --blame some/file.txt
```

- Specific line at file

```
$ git brws --blame some/file.txt#L5
```

- Specific range at file

```
$ git brws --blame some/file.txt#L5-L9
```

### Cooperate with other tools

With `-u` option, `git brws` outputs URL to stdout.

For example, in Vim, you can write your repository URL to your text instantly.

```
:r!git brws -u
```

And below can open editing file in your browser.

```
:!git brws %
```

## Customization

Some environment variables are available to customize behavior of `git-brws`.

| Variable | Description |
|----------|-------------|
| `$GIT_BRWS_GIT_COMMAND` | Git command to use. If not specified, `"git"` will be used. |
| `$GIT_BRWS_GHE_URL_HOST` | When you use your own GitHub Enterprise repository, you can specify its host to this variable. By default, `git brws` detects `^github\.` as GHE host. If your GHE repository host does not match it, please specify this variable. If your repository is `https://example-repo.org/user/repo`, `example-repo.org` should be set. |
| `$GIT_BRWS_GHE_SSH_PORT` | When you set a number to it, the number will be used for the ssh port for GitHub Enterprise URLs. |
| `$GIT_BRWS_GITLAB_URL_HOST` | When you use self-hosted GitLab instance, you can specify its host to this variable. By default, `git brws` detects host matching to `^gitlab\.` as GitLab. If your GitLab URL host does not match it, please specify this variable. If your repository is hosted at `https://your-code.net/user/repo`, `your-code.net` should be set. |
| `$GIT_BRWS_GITLAB_SSH_PORT` | When you set a number to it, the number will be used for the ssh port for self-hosted GitLab URLs. This is useful when your environment hosts GitLab to non-trivial ssh port URL. |
| `$GIT_BRWS_GITHUB_TOKEN` | This variable is used for `--pr` (or `-p`) only. API access token for github.com. They are optional, but useful for avoiding API rate limit and accessing to private repositories. Please generate a token from https://github.com/settings/tokens/new |
| `$GITHUB_TOKEN` | Ditto. When `GIT_BRWS_GITHUB_TOKEN` is not set, `GITHUB_TOKEN` is looked. |
| `$GIT_BRWS_GHE_TOKEN` | This variable is used for `--pr` (or `-p`) only. API access token for GitHub Enterprise instance. It is sometimes mandatory (depending on your GHE instance configuration). Please generate a token from `https://{YOUR GHE HOST}/settings/tokens/new`. |
| `$GIT_BRWS_BROWSE_COMMAND` | Command to open URL. If this value is specified, the command is executed with URL as first argument to browse the URL. |
| `$GIT_BRWS_SHORT_COMMIT_HASH` | Setting `true` will use 7-letters short commit hash like `78fbce6` for URLs. |
| `$https_proxy` | This variable is used for `--pr` (or `-p`) only. An HTTPS Proxy server URL if you use a web proxy. |

## Related Projects

- [hub (`hub browse`)](https://hub.github.com/)
- [git-open](https://github.com/paulirish/git-open)
- [open-browser-github.vim](https://github.com/tyru/open-browser-github.vim)
- [git-browse](https://github.com/albertyw/git-browse)

## TODOs

Please see [the project page](https://github.com/rhysd/git-brws/projects/1).

## Development

Please see [CONTRIBUTING.md](https://github.com/rhysd/git-brws/blob/master/CONTRIBUTING.md).

## License

Distributed under [the MIT license](LICENSE.txt).

[crate-badge]: https://img.shields.io/crates/v/git-brws.svg
[crate]: https://crates.io/crates/git-brws
[travis-ci]: https://travis-ci.org/rhysd/git-brws
[travis-badge]: https://travis-ci.org/rhysd/git-brws.svg?branch=master
[appveyor-badge]: https://ci.appveyor.com/api/projects/status/q9gvpd30k1k5jsf0/branch/master?svg=true
[appveyor]: https://ci.appveyor.com/project/rhysd/git-brws/branch/master
[codecov-badge]: https://codecov.io/gh/rhysd/git-brws/branch/master/graph/badge.svg
[codecov]: https://codecov.io/gh/rhysd/git-brws
[release page]: https://github.com/rhysd/git-brws/releases
[gitlab-pages]: https://docs.gitlab.com/ee/user/project/pages/getting_started_part_one.html#project-websites
[bitbucket-cloud]: https://confluence.atlassian.com/bitbucket/publishing-a-website-on-bitbucket-cloud-221449776.html


CHANGELOG.md
<a name="v0.11.12"></a>
# [v0.11.12](https://github.com/rhysd/git-brws/releases/tag/v0.11.12) - 18 Sep 2020

- **Fix:** Support [GitLab's subgroups](https://docs.gitlab.com/ee/user/group/subgroups/) (#28)
- **Improve:** Check repository name and user/org name more strictly
- **Improve:** Update crate dependencies

[Changes][v0.11.12]


<a name="v0.11.11"></a>
# [v0.11.11](https://github.com/rhysd/git-brws/releases/tag/v0.11.11) - 03 Sep 2020

- **New:** Add `--current-branch` option (`-c` in short option) to open current branch in browser. For example, when `git brws` opens `https://github.com/foo/bar` and current branch is `some-branch`, `git brws -c` opens `https://github.com/foo/bar/tree/some-branch`.
- **Fix:** Handle invalid empty branch name at `-b` option correctly (#26)
- **Fix:** URL is not correct when a directory path is given (#25)
- **Improve:** Update dependent crates

[Changes][v0.11.11]


<a name="v0.11.10"></a>
# [v0.11.10](https://github.com/rhysd/git-brws/releases/tag/v0.11.10) - 05 Aug 2020

- **Fix:** Parsing SCP-like syntax `user@server:user/repo.git` failed when `user` starts with number value (#24)
- **Fix:** Only `git@` was considered as user of above SCP-like syntax. Actually `vs-ssh.visualstudio.com` service provides user names other than `git@`

[Changes][v0.11.10]


<a name="v0.11.9"></a>
# [v0.11.9](https://github.com/rhysd/git-brws/releases/tag/v0.11.9) - 27 Apr 2020

- **New:** Add `$GIT_BRWS_SHORT_COMMIT_HASH` environment variable. When this value is set, `git-brws` will use 7-letters short commit hash like `78fbce6` for URLs. ([#22](https://github.com/rhysd/git-brws/issues/22))
- **Improve:** Update dependencies

[Changes][v0.11.9]


<a name="v0.11.8"></a>
# [v0.11.8](https://github.com/rhysd/git-brws/releases/tag/v0.11.8) - 11 Apr 2020

- **New:** Add `$GIT_BRWS_GITLAB_URL_HOST` to specify URL host for self-hosted GitLab instance. Please see README.md or `man` manual for more details
- **Fix:** Windows build was broken
- **Improve:** Better error message

[Changes][v0.11.8]


<a name="v0.11.7"></a>
# [v0.11.7](https://github.com/rhysd/git-brws/releases/tag/v0.11.7) - 08 Feb 2020

- **Improve:** `-C` is used instead of `--git-dir` for executing `git` commands to avoid weird gitdir issue on Windows
- **Improve:** `?expand=1` is added to 'Create pull request' page when opening a repository with `--pr` and a pull request is not yet created. It opens a text form for pull request description automatically
- **Improve:** Dependencies were updated
- **Fix:** Setting empty string to `$https_proxy` did not work
- **Dev:** CI for Windows was fixed. It failed due to weird reason
- **Dev:** Use `std::error::Error` trait to handle nested errors

[Changes][v0.11.7]


<a name="v0.11.6"></a>
# [v0.11.6](https://github.com/rhysd/git-brws/releases/tag/v0.11.6) - 16 Jan 2020

- **Improve:** `reqwest` crate was updated to 0.10
  - Number of dependencies and build time were 1.5x reduced
  - Binary size gets 1.3~1.4x smaller
- **Improve:** `--pr` is now faster by sending multiple GitHub API requests at the same time. Thanks to the `async-await` feature of latest Rust

[Changes][v0.11.6]


<a name="0.11.5"></a>
# [0.11.5](https://github.com/rhysd/git-brws/releases/tag/0.11.5) - 21 Nov 2019

- **Fix:** Opening file or directory is slow when remote repository has so many (1000~) branches (#19)
- **Improve:** More accurately convert abbreviated notation of SSH URL

[Changes][0.11.5]


<a name="0.11.3"></a>
# [0.11.3](https://github.com/rhysd/git-brws/releases/tag/0.11.3) - 04 Nov 2019

- **New:** Repositories hosted on [Azure DevOps](https://azure.microsoft.com/ja-jp/services/devops/) are now supported. Special thanks to @bgianfo for adding this. (#15)
- **Improve:** On WSL, [`wslview`](https://github.com/wslutilities/wslu) may be used to open a page in browser if available. This was added by updating `open` crate from 0.12 to 0.13.
- **Improve:** Update dependencies (especially `url` crate had major update)

[Changes][0.11.3]


<a name="0.11.2"></a>
# [0.11.2](https://github.com/rhysd/git-brws/releases/tag/0.11.2) - 19 May 2019

- **Fix:** URL is wrong when opening 'New Pull Request' page of parent repository (#13)
- **Improve:** Reduce binary size (about 30% smaller) by tweaking build configuration

[Changes][0.11.2]


<a name="0.11.1"></a>
# [0.11.1](https://github.com/rhysd/git-brws/releases/tag/0.11.1) - 30 Apr 2019

**Fix:** Cause an error when attempting to blame directory path

[Changes][0.11.1]


<a name="0.11.0"></a>
# [0.11.0](https://github.com/rhysd/git-brws/releases/tag/0.11.0) - 29 Apr 2019

- **Change:** Open 'Create Pull Request' page when pull request is not created yet for the branch. Previous behavior caused an error in the case.
- **New:** Support 'Blame' page for file path with `--blame` or `-B` option. Not only file path, with line such as `/path#L123` and with range such as `/path#L1-2` are supported. This feature is available for all services; GitHub, GitHub:Enterprise, GitLab and bitbucket. #11
- **New:** `$GIT_BRWS_BROWSE_COMMAND` environment variable was introduced to customize how to open the URL. The command specified with the environment variable is invoked with URL as the first argument for opening the URL.
- **Fix:** When opening file path, fallback into branch name instead of commit hash if the latest commit is not contained in remote branch. Since the commit page where the commit is not contained in remote is 404. #12

[Changes][0.11.0]


<a name="0.10.1"></a>
# [0.10.1](https://github.com/rhysd/git-brws/releases/tag/0.10.1) - 12 Feb 2019

- **New:** Argument now accepts [project's tag page](https://github.com/rhysd/git-brws/tree/0.10.0). GitHub, GitHub Enterprise, GitLab and Bitbucket are supported
- **Fix:** `$https_proxy` and `$GITHUB_TOKEN` were not referred
- Improve error messages when argument cannot be handled and Git command cannot run

[Changes][0.10.1]


<a name="0.10.0"></a>

src/argv.rs
use crate::async_runtime;
use crate::config::{Config, EnvConfig};
use crate::error::{Error, ErrorKind, Result};
use crate::git::Git;
use crate::github_api::Client;
use getopts::Options;
use std::env;
use std::ffi::OsStr;
use std::fs;
use std::path::PathBuf;

fn handle_scp_like_syntax(mut url: String) -> String {
    // ref: https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols

    if url.contains("://") {
        // When url is a URL like https://server/project.git
        return url;
    }

    // When the target is an scp-like syntax: [user@]server:project.git
    // Handle ':' in the syntax. Note that GitHub user name may start with number (#24)
    if let Some(i) = url.find(':') {
        let after_colon = i + 1;
        if let Some(i) = url[after_colon..].find(':') {
            // When port number exists
            //  git@service.com:123:user/repo.git -> git@service.com:123/user/repo.git
            let i = after_colon + i;
            url.replace_range(i..i + 1, "/");
        } else {
            // When a port number is omitted, default SSH port is 22
            //  git@service.com:user/repo.git -> git@service.com:22/user/repo.git
            url.insert_str(after_colon, "22/"); // Insert port number after colon
        }
    }

    // Examples:
    //  git@service.com:22/user/repo.git -> ssh://git@service.com:22/user/repo.git
    url.insert_str(0, "ssh://");
    url
}

#[cfg_attr(feature = "cargo-clippy", allow(clippy::large_enum_variant))]
#[derive(Debug)]
pub enum Parsed {
    Help(String),
    Version(&'static str),
    OpenPage(Config),
}

fn is_scp_like_syntax_with_user(s: &str) -> bool {
    // SCP-like syntax like user@project:project
    // user@ cannot be omitted in SCP-like syntax because `s` can be a search query
    //
    // Note that checking it starts with "git@" does not work because SSH protocol for
    // visualstudio.com is "mycompany@vs-ssh.visualstudio.com:v3/project.git"
    if s.contains(char::is_whitespace) {
        return false; // Containing whitespace means it's a search query
    }
    for (i, c) in s.char_indices() {
        match c {
            '@' if i == 0 => return false,          // Seems query like "@rhysd"
            '@' => return s[i + 1..].contains(':'), // SCP-like syntax must also contain ':'
            ':' => return false,                    // '@' must be put before ':'
            _ => continue,
        }
    }
    false
}

fn normalize_repo_format(mut slug: String, env: &EnvConfig) -> Result<String> {
    if slug.is_empty() {
        return Error::err(ErrorKind::BrokenRepoFormat { input: slug });
    }

    // - URL like https://server/project
    // - SCP-like syntax user@project:project
    //
    // Note: user@ cannot be omitted in SCP-like syntax because `slug` can be search word
    if slug.starts_with("https://")
        || slug.starts_with("ssh://")
        || slug.starts_with("http://")
        || slug.starts_with("file://")
        || is_scp_like_syntax_with_user(&slug)
    {
        if !slug.ends_with(".git") {
            slug.push_str(".git");
        }
        return Ok(slug);
    }

    match slug.chars().filter(|c| *c == '/').count() {
        1 => Ok(format!("https://github.com/{}.git", slug)),
        2 => Ok(format!("https://{}.git", slug)),
        0 => {
            let client = Client::build("api.github.com", &env.github_token, &env.https_proxy)?;
            async_runtime::blocking(client.most_popular_repo_by_name(&slug))
                .map(|repo| repo.clone_url)
        }
        _ => Error::err(ErrorKind::BrokenRepoFormat { input: slug }),
    }
}

fn get_cwd(specified: Option<String>) -> Result<PathBuf> {
    if let Some(dir) = specified {
        let p = fs::canonicalize(&dir)?;
        if !p.exists() {
            return Error::err(ErrorKind::SpecifiedDirNotExist { dir });
        }
        Ok(p)
    } else {
        Ok(env::current_dir()?.canonicalize()?)
    }
}

const USAGE: &str = "\
Usage: git brws [Options] {Args}

  Open a repository, file, commit, diff or pull request, issue or project's
  website in your web browser from command line.
  GitHub, Bitbucket, GitLab, GitHub Enterprise, Azure DevOps are supported as
  hosting service.
  git-brws looks some environment variables for configuration. Please see
  https://github.com/rhysd/git-brws#readme for more details.

Examples:
  - Current repository:

    $ git brws

  - GitHub repository:

    $ git brws -r rhysd/git-brws

  - Most popular GitHub repository by name:

    $ git brws -r git-brws

  - File:

    $ git brws some/file.txt

  - Commit:

    $ git brws HEAD~3

  - Tag:

    $ git brws 0.10.0

  - Diff between commits:

    $ git brws HEAD~3..HEAD

  - Diff between topic and topic's merge base commit:

    $ git brws master...topic

  - Line 123 of file:

    $ git brws some/file.txt#L123

  - Range from line 123 to line 126 of file:

    $ git brws some/file.txt#L123-L126

  - Pull request page (for GitHub and GitHub Enterprise):

    $ git brws --pr

  - Website of repository at current directory

    $ git brws --website

  - Website of other repository

    $ git brws --website --repo react

  - Issue page:

    $ git brws '#8'";

impl Parsed {
    pub fn parse_iter<I>(argv: I) -> Result<Parsed>
    where
        I: IntoIterator,
        I::Item: AsRef<OsStr>,
    {
        let mut opts = Options::new();

        opts.optopt("r", "repo", "Shorthand format (repo, user/repo, host/user/repo) or Git URL you want to see. When only repo name is specified, most popular repository will be searched from GitHub", "REPO");
        opts.optopt("b", "branch", "Branch name to browse", "BRANCH");
        opts.optopt(
            "d",
            "dir",
            "Directory path to the repository. Default value is current working directory",
            "PATH",
        );
        opts.optopt("R", "remote", "Remote name (e.g. origin). Default value is a remote the current branch is tracking. If current branch tracks no branch, it falls back to 'origin'", "REMOTE");
        opts.optflag(
            "u",
            "url",
            "Output URL to stdout instead of opening in browser",
        );
        opts.optflag(
            "p",
            "pr",
            "Open pull request page instead of repository page. If not existing, open 'Create Pull Request' page",
        );
        opts.optflag(
            "w",
            "website",
            "Open website page instead of repository page (homepage URL for GitHub, GitLab pages, Bitbucket Cloud)",
        );
        opts.optflag(
            "B",
            "blame",
            "Open blame page instead of repository page. File path to blame must be passed also.",
        );
        opts.optflag(
            "c",
            "current-branch",
            "Open the current branch instead of default branch",
        );
        opts.optflag("h", "help", "Print this help");
        opts.optflag("v", "version", "Show version");

        let matches = opts.parse(argv.into_iter().skip(1))?;

        if matches.opt_present("h") {
            return Ok(Parsed::Help(opts.usage(USAGE)));
        }

        if matches.opt_present("v") {
            return Ok(Parsed::Version(
                option_env!("CARGO_PKG_VERSION").unwrap_or("unknown"),
            ));
        }

        let env = EnvConfig::from_iter(env::vars())?.with_global_env();
        let cwd = get_cwd(matches.opt_str("d"))?;
        let git = Git::new(&cwd, &env.git_command);
        let branch = if let Some(b) = matches.opt_str("b") {
            if b.is_empty() {
                return Error::err(ErrorKind::BranchNameEmpty);
            }
            Some(b)
        } else if matches.opt_present("c") {
            Some(git.current_branch()?)
        } else {
            None
        };
        let (repo_url, remote) = match (matches.opt_str("r"), matches.opt_str("R")) {
            (Some(repo), remote) => {
                if !matches.free.is_empty() {
                    return Error::err(ErrorKind::ArgsNotAllowed {
                        flag: "--repo {repo}",
                        args: matches.free,
                    });
                }
                (normalize_repo_format(repo, &env)?, remote)
            }
            (None, remote) => {
                let (url, remote) = if let Some(remote) = remote {
                    (git.remote_url(&remote)?, remote)
                } else {
                    git.tracking_remote_url(&branch)?
                };
                (url, Some(remote))
            }
        };

        let repo_url = handle_scp_like_syntax(repo_url);

        Ok(Parsed::OpenPage(Config {
            repo_url,
            branch,
            cwd,
            stdout: matches.opt_present("u"),
            pull_request: matches.opt_present("p"),
            website: matches.opt_present("w"),
            blame: matches.opt_present("B"),
            args: matches.free,
            remote,
            env,
        }))
    }
}
